cmd/logger/http.go

```diff
@@ -0,0 +1,89 @@
+/*
+ * Minio Cloud Storage, (C) 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package logger
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"net/http"
+)
+
+// HTTPTarget implements loggerTarget and sends the json
+// format of a log entry to the configured http endpoint.
+// An internal buffer of logs is maintained but when the
+// buffer is full, new logs are just ignored and an error
+// is returned to the caller.
+type HTTPTarget struct {
+	// Channel of log entries
+	logCh chan logEntry
+	// HTTP(s) endpoint
+	endpoint  string
+	transport *http.Transport
+}
+
+func (h *HTTPTarget) startHTTPLogger() {
+	// Create a routine which sends json logs received
+	// from an internal channel.
+	go func() {
+		for entry := range h.logCh {
+			logJSON, err := json.Marshal(&entry)
+			if err != nil {
+				continue
+			}
+
+			req, err := http.NewRequest("POST", h.endpoint, bytes.NewBuffer(logJSON))
+			req.Header.Set("Content-Type", "application/json")
+
+			client := &http.Client{
+				Transport: h.transport,
+			}
+			resp, err := client.Do(req)
+			if err != nil {
+				continue
+			}
+			if resp.Body != nil {
+				resp.Body.Close()
+			}
+		}
+	}()
+}
+
+// NewHTTP initializes a new logger target which
+// sends log over http to the specified endpoint
+func NewHTTP(endpoint string, transport *http.Transport) LoggingTarget {
+	h := HTTPTarget{
+		endpoint:  endpoint,
+		transport: transport,
+		logCh:     make(chan logEntry, 10000),
```

Comment:
If we place the `http.Client` here, and re-use, I think connection re-use would be automatic?

Reply:
> We should re-use the connection across log messages for better efficiency.

Connection reusing means establishing a permanent connection with the server ? if yes, then I don't recommend it since an established connection will consume one tcp port meanwhile servers are not supposed to send logs regularly. 

What do you think ?

Reply:
The idle connection timeout of 30 seconds takes care of this. If multiple logs were generated, then repeatedly opening TCP connections, and negotiating TLS handshake, etc is too repetitive - it is better to re-use an established connection.

Reply:
> The idle connection timeout of 30 seconds takes care of this. If multiple logs were generated, then repeatedly opening TCP connections, and negotiating TLS handshake, etc is too repetitive - it is better to re-use an established connection.

Yeah, didn't know about that 30 seconds timeout, fixed now